Show the full URL in a timeout error

giles-v opened the pull request:
For context, please see https://github.com/CondeNast/copilot-util/issues/9.

*tl;dr*: timeout errors in this library only return the request `path`, which can make it difficult to figure out which host was being queried. This PR proposes that we log the full URL which timed out.

paulfryzel commented:
Also shouldn't this apply to all Error types in the SDK as well?

=> https://github.com/CondeNast/copilot-util/tree/master/lib/error#error_http

pgoldrbx commented:
Right - all error types would be ideal. 

It looks like codeship might not be connected properly - we can try and fix that.

Also - @giles-v we should probably try and add a test for this change if possible

pgoldrbx commented on lib/http/response/index.js:
```diff
@@ -23,7 +24,7 @@ module.exports = function onResponse(res) {
       res.destroy();
       var he = new error.HttpError(statusCode);
-      he.url = res.req && res.req.path;
+      he.url = res.req && url.format(res.req);
```
you can add a test for this I think
https://github.com/CondeNast/copilot-util/blob/master/test/test-http-response.js#L51

pgoldrbx replied:
So, this will definitely not work. `res.req` is an instance of `http.ClientRequest`. The URL at this point is not easily recoverable and it may be ill advised to try and build it from the request object. I will dig a bit more.

pgoldrbx replied:
we _could_ use `res.headers.location` perhaps?

pgoldrbx replied:
@paulfryzel do you have any thoughts here?

brokentone replied:
I don't think `res.headers.location` is a good approach

giles-v commented:
@paulfryzel @pgoldrbx I've added code for the other invocations in the lib. But the test is going to take me longer. I had a few cracks at faking a timeout without success, and we'll need one to mock 500 to throw the one in the `response` too. 

I'll dig more into this later this week.

pgoldrbx commented:
@giles-v @brokentone @paulfryzel I'd like to revisit this work. Any thoughts?

giles-v commented:
So this kind of ground to a halt in two areas: my testing issues (should be easily resolvable) and this comment of yours, which is a little less black and white:

> The URL at this point is not easily recoverable and it may be ill advised to try and build it from the request object

My sense is that it's not _that_ ill-advised; we don't need absolute precision here and if we can read off `request.protocol` and `request.host` that's usually going to be all we need to solve the core debugging issue. 

But at this point I was vaguely wondering whether we'd be better advised to lean towards an [X-Ray](https://aws.amazon.com/xray/) type solution to this problem rather than following the initial route on this PR. Is that too wide / far-off a target?

paulfryzel commented:
@giles-v I'm fine if we want to come back to this with a direct option for the host+procotol. I don't think we should invest too much more in tracing solutions for this library until we talk about its longer-term existence with other options such as `node-fetch`.

giles-v commented:
@pgoldrbx @paulfryzel I've adjusted this slightly to a reasonably safe and useful route, and added tests.

Followup review would be appreciated! Sorry for the delay...

scherrymomin commented:
Hey guys as much as I love getting git notifications I’m still on this account. Could someone remove me please? Hope y’all are well.

scherrymomin commented:
Cc @paulfryzel

giles-v commented:
Hey there @scherrymomin! This is a public repo, you can probably stop the notifications by simply unwatching it top-right. Let me know if that doesn't work.

giles-v commented:
Just a bump, still looking for review @pgoldrbx @paulfryzel
